include/albatross/src/cereal/suitesparse.hpp

```diff
@@ -0,0 +1,486 @@
+                         std::size_t expected_bytes) {
+  ALBATROSS_ASSERT(nullptr != destination);
+  std::string payload;
+  ar(::cereal::make_nvp(name, payload));
+  if (::cereal::traits::is_text_archive<Archive>::value) {
+    payload = base64::decode(payload);
+  }
+  const std::string decompressed =
+      gzip::decompress(payload.data(), payload.size());
+  ALBATROSS_ASSERT(decompressed.size() == expected_bytes);
+  std::memcpy(destination, decompressed.data(), decompressed.size());
+}
+
+// This exists because 1) the suitesparse memory functions require a
+// cholmod_common pointer for statistics 2) the suitesparse realloc
+// helpers have difficult state properties.
+inline void suitesparse_cereal_realloc(void **p, std::size_t count,
+                                       std::size_t elem_size) {
+  if (nullptr != *p) {
+    free(*p);
+  }
+  *p = malloc(count * elem_size);
+  ALBATROSS_ASSERT(nullptr != *p);
+}
+
+}  // namespace detail
+
+template <class Archive>
+inline void save(Archive &ar, cholmod_common const &cc,
+                 std::uint32_t version ALBATROSS_UNUSED) {
+  ar(CEREAL_NVP(cc.dbound));
+  ar(CEREAL_NVP(cc.grow0));
+  ar(CEREAL_NVP(cc.grow1));
+  ar(CEREAL_NVP(cc.grow2));
+  ar(CEREAL_NVP(cc.maxrank));
+  ar(CEREAL_NVP(cc.supernodal_switch));
+  ar(CEREAL_NVP(cc.supernodal));
+  ar(CEREAL_NVP(cc.final_asis));
+  ar(CEREAL_NVP(cc.final_super));
+  ar(CEREAL_NVP(cc.final_ll));
+  ar(CEREAL_NVP(cc.final_pack));
+  ar(CEREAL_NVP(cc.final_monotonic));
+  ar(CEREAL_NVP(cc.final_resymbol));
+  static_assert(sizeof(cc.zrelax) == 3 * sizeof(double),
+                "cholmod_common::zrelax expected to have 3 elements.");
+  ar(::cereal::make_nvp("cc.zrelax0", cc.zrelax[0]));
+  ar(::cereal::make_nvp("cc.zrelax1", cc.zrelax[1]));
+  ar(::cereal::make_nvp("cc.zrelax2", cc.zrelax[2]));
+  static_assert(sizeof(cc.nrelax) == 3 * sizeof(std::size_t),
+                "cholmod_common::nrelax expected to have 3 elements.");
+  ar(::cereal::make_nvp("cc.nrelax0", cc.nrelax[0]));
+  ar(::cereal::make_nvp("cc.nrelax1", cc.nrelax[1]));
+  ar(::cereal::make_nvp("cc.nrelax2", cc.nrelax[2]));
+  ar(CEREAL_NVP(cc.prefer_zomplex));
+  ar(CEREAL_NVP(cc.prefer_upper));
+  ar(CEREAL_NVP(cc.quick_return_if_not_posdef));
+  ar(CEREAL_NVP(cc.prefer_binary));
+  ar(CEREAL_NVP(cc.print));
+  ar(CEREAL_NVP(cc.precise));
+  ar(CEREAL_NVP(cc.try_catch));
+  // Do not serialise the user error handler.
+  ar(CEREAL_NVP(cc.nmethods));
+  ar(CEREAL_NVP(cc.current));
+  ar(CEREAL_NVP(cc.selected));
+  constexpr std::size_t max_methods = sizeof(cc.method) / sizeof(cc.method[0]);
+  for (std::size_t m = 0; m < max_methods; ++m) {
+    ar(cc.method[m]);
+  }
+  // method
+  ar(CEREAL_NVP(cc.postorder));
+  ar(CEREAL_NVP(cc.default_nesdis));
+  ar(CEREAL_NVP(cc.metis_memory));
+  ar(CEREAL_NVP(cc.metis_dswitch));
+  ar(CEREAL_NVP(cc.metis_nswitch));
+  ar(::cereal::make_nvp("nrow", cc.nrow));
```

Comment:
why does it sometimes use `CEREAL_NVP` and othertimes explicit names?

Reply:
In most of the occurrences, this happens because I'm serialising elements of a fixed-length array.

The reason it happens this way for `nrow`, `iworksize` and `xworksize` is that I want to deserialise them to stack variables so I can calculate allocation sizes etc., rather than deserialising them to elements of `cholmod_common` directly (because initialisation functions can and do modify various elements of `cholmod_common` structures).  I'll add a comment here.